lib/3scale/backend/storage_async/client.rb

```
@@ -41,131 +33,17 @@ def instance(reset = false)
         end
 
         def initialize(opts)
-          host, port = opts[:url].match(HOST_PORT_REGEX).captures if opts[:url]
-          host ||= DEFAULT_HOST
-          port ||= DEFAULT_PORT
-
-          endpoint = Async::IO::Endpoint.tcp(host, port)
-          @redis_async = Async::Redis::Client.new(
-            endpoint, limit: opts[:max_connections]
-          )
-          @building_pipeline = false
+          @redis_async = initialize_client(opts)
```

Actually I disagree. Either this needs to initialize a client per thread (making it thread safe), or `::instance` should return a different objects per thread (preferable), or `::instance` should fail if it is ever called from another thread, effectively "failing fast". For example memoize thread that initially called it and err in case another thread calls it ever.

---

Presently we can never be sure that that this would not be used by multiple threads, relying on deployment options without any check is unreliable and prone to failure.